src/components/application_manager/src/application_manager_impl.cc

```diff
@@ -2234,6 +2287,60 @@ bool ApplicationManagerImpl::ConvertSOtoMessage(
   return true;
 }
 
+MessageValidationResult ApplicationManagerImpl::ValidateMessageBySchema(
+    const Message& message) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  smart_objects::SmartObject so;
+  using namespace protocol_handler;
+  switch (message.protocol_version()) {
+    case MajorProtocolVersion::PROTOCOL_VERSION_5:
+    case MajorProtocolVersion::PROTOCOL_VERSION_4:
+    case MajorProtocolVersion::PROTOCOL_VERSION_3:
+    case MajorProtocolVersion::PROTOCOL_VERSION_2: {
+      const bool conversion_result =
+          formatters::CFormatterJsonSDLRPCv2::fromString(
+              message.json_message(),
+              so,
+              message.function_id(),
+              message.type(),
+              message.correlation_id());
+      if (!conversion_result) {
+        return INVALID_JSON;
+      }
+
+      if (!mobile_so_factory().attachSchema(so, true)) {
+        return INVALID_METADATA;
+      }
+
+      if (so.validate() != smart_objects::Errors::OK) {
+        return SCHEMA_MISMATCH;
+      }
+      break;
+    }
+    case MajorProtocolVersion::PROTOCOL_VERSION_HMI: {
+      const int32_t conversion_result = formatters::FormatterJsonRpc::
+          FromString<hmi_apis::FunctionID::eType, hmi_apis::messageType::eType>(
+              message.json_message(), so);
+      if (0 != conversion_result) {
+        LOG4CXX_WARN(logger_,
+                     "Failed to parse json from HMI: " << conversion_result);
+        return INVALID_JSON;
```

Comment:
@JackLivio to check result of json processing. Different error codes cause different behavior.